# Import Libraries

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import mnist

# Load MNIST

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Reshape

In [3]:
x_train = x_train.reshape(-1, 28, 28, 1).astype("float32") / 255
x_test  = x_test.reshape(-1, 28, 28, 1).astype("float32") / 255

# Build CNN Model

In [4]:
model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),
    
    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    
    layers.Conv2D(64, (3,3), activation='relu'),

    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

c:\Users\ArkaT\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


# Compile

In [5]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train

In [6]:
history = model.fit(
    x_train, y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 8s 8ms/step - accuracy: 0.9458 - loss: 0.1771 - val_accuracy: 0.9867 - val_loss: 0.0470
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9847 - loss: 0.0503 - val_accuracy: 0.9883 - val_loss: 0.0436
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.9887 - loss: 0.0357 - val_accuracy: 0.9905 - val_loss: 0.0337
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.9912 - loss: 0.0269 - val_accuracy: 0.9867 - val_loss: 0.0500
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9932 - loss: 0.0217 - val_accuracy: 0.9925 - val_loss: 0.0318


# Evaluate

In [8]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9905 - loss: 0.0282
Test Accuracy: 0.9904999732971191


The MNIST CNN experiment delivered outstanding results, reaching a training accuracy of 99.05% with a very low loss of 0.0282. The test accuracy is equally high at 99.05%, showing that the model can reliably recognize handwritten digits with exceptional precision. The low loss value indicates strong confidence in its predictions, while the nearly identical training and testing accuracy suggests excellent generalization with no significant overfitting. Overall, these results confirm that the CNN architecture combining convolutional, pooling, and dense layers is highly effective for digit classification and successfully learns the key features needed to distinguish all ten MNIST digit classes.